src/main/java/gregtech/client/particle/GTParticleManager.java

```diff
@@ -18,165 +20,178 @@
- * @Description: ParticleManger register, spawn, efficient rendering, update our custom particles.
+ * Singleton class responsible for managing, updating and rendering {@link GTParticle} instances.
  */
 @SideOnly(Side.CLIENT)
 public class GTParticleManager {
-    public final static GTParticleManager INSTANCE = new GTParticleManager();
 
+    public static final GTParticleManager INSTANCE = new GTParticleManager();
+
+    @Nullable
     private static World currentWorld = null;
-    private static final Minecraft mc = Minecraft.getMinecraft();
 
-    private final Map<IGTParticleHandler, ArrayDeque<GTParticle>> renderQueueBack = new HashMap<>();
-    private final Map<IGTParticleHandler, ArrayDeque<GTParticle>> renderQueueFront = new HashMap<>();
-    private final Queue<Tuple<IGTParticleHandler, GTParticle>> newParticleQueue = new ArrayDeque<>();
+    private final Map<IRenderSetup, ArrayDeque<GTParticle>> depthEnabledParticles = new Object2ObjectLinkedOpenHashMap<>();
+    private final Map<IRenderSetup, ArrayDeque<GTParticle>> depthDisabledParticles = new Object2ObjectLinkedOpenHashMap<>();
 
-    public void addEffect(GTParticle... particles) {
-        for (GTParticle particle : particles) {
-            if (particle.getGLHandler() != null) {
-                newParticleQueue.add(new Tuple<>(particle.getGLHandler(), particle));
-            } 
-        }
+    private final List<GTParticle> newParticleQueue = new ArrayList<>();
+
+    public void addEffect(@Nonnull GTParticle particles) {
+        newParticleQueue.add(particles);
     }
 
     public void updateEffects() {
-        updateEffectLayer();
+        if (!depthEnabledParticles.isEmpty()) {
+            updateQueue(depthEnabledParticles);
+        }
+        if (!depthDisabledParticles.isEmpty()) {
+            updateQueue(depthDisabledParticles);
+        }
         if (!newParticleQueue.isEmpty()) {
-            for (Tuple<IGTParticleHandler, GTParticle> handlerParticle = newParticleQueue.poll(); handlerParticle != null; handlerParticle = newParticleQueue.poll()) {
-                IGTParticleHandler handler = handlerParticle.getFirst();
-                GTParticle particle = handlerParticle.getSecond();
-                Map<IGTParticleHandler, ArrayDeque<GTParticle>> renderQueue = particle.getFXLayer() > 0 ? renderQueueFront : renderQueueBack;
-                if (!renderQueue.containsKey(handler)) {
-                    renderQueue.put(handler, new ArrayDeque<>());
-                }
-                ArrayDeque<GTParticle> arrayDeque = renderQueue.get(handler);
-                if (arrayDeque.size() > 6000) {
-                    arrayDeque.removeFirst().setExpired();
+            for (GTParticle particle : newParticleQueue) {
+                var queue = particle.shouldDisableDepth() ? depthDisabledParticles : depthEnabledParticles;
+
+                ArrayDeque<GTParticle> particles = queue.computeIfAbsent(particle.getRenderSetup(), setup -> new ArrayDeque<>());
```

Comment:
Since `particle.getRenderSetup()` is nullable, we may we wish to skip this operation altogether. Associating a deque for a null key sounds like cause for potential issues here. If using null this way is intentional, I think some sort of explicit dummy `IRenderSetup` would be more appropriate.

Reply:
Nullable render setup instance is an explicit decision here, because I found the notion of null render setups more straightforward - "no render setup" directly correlates to "no render setup instance (null)", while using nonnull representation you have to say there should be _something_ even if the thing that object represents is just _nothing_. Plus it avoids pitfall of users just making empty impl for the interface because returning null crashes the game with NPE, breaking render setup's equality contract in the process.

The inner part of the API (the part of code this conversation is taking place) also uses nullable keys, because (1) the collection lib gracefully supports it (fuck you stdlib), (2) it's (probably) (slightly) faster and (3) there's no need to make null-object representation if base API does not uses it.

While the key nullability is undocumented, the render setup is used in one place, and the collection is only internal, so I think there's no big risk there. I personally would have liked nullable annotation in the type (like `Map<@Nullable IRenderSetup, ArrayDeque<GTParticle>>`), but I couldn't do that since I was using findbugs nullable annotations, which cannot target types.

Still I probably have to document the nullability lol